### PR TITLE
for python 3.8-3.12

### PR DESCRIPTION
**Pull request by azazellochg**
Related to https://github.com/scipion-em/scipion-pyworkflow/issues/476

### CODE REVIEW

**Comment by pconesa** [comment]
hi! You have removed the conditional ones per python version.

This will force user with those envs in python3.6 to become corrupted. I think we should keep them until we decide/think how to migrate users. If we decide to do so or just ask them to reinstall.

**Comment by pconesa** [comment]
Yes, but for basic and simple dependencies. When it comes to "cascade 
dependencies" it could fail.

**Comment by azazellochg** [comment]
Understood, then we will need to figure them out for all pythons from 3.6
to 3.11
Best regards,
Grigory

--------------------------------------------------------------------------------
Grigory Sharov, Ph.D.

MRC Laboratory of Molecular Biology,
Francis Crick Avenue,
Cambridge Biomedical Campus,
Cambridge CB2 0QH, UK.
tel. +44 (0) 1223 267228 <+44%201223%20267228>
e-mail: ***@***.***

**Comment by azazellochg** [comment]
I'll put back the conditions tomorrow

**Comment by azazellochg** [comment]
Builds 3.6 to 3.12 now work: https://github.com/scipion-em/scipion-pyworkflow/actions/runs/6797546197/job/18480094449 
3.12  requires newer scipy than scipion-em-xmipp allows at the moment
I will try compiling xmipp with these new deps and post here

**Comment by azazellochg** [comment]
The main showstopper is xmipp. It is difficult to push updates to their requirements. I'm working on compiling it with newer GCC and also these new requirements and make a thread on discord with results. There are lots of things to check  like matplotlib, numpy/scipy, scikit etc. I've got a new laptop and will install this new scipion and begin testing. Hopefully by next spring we can push this PR.

**Comment by azazellochg** [comment]
good question. A test would be to install all from scratch from devel branch with 3.8 and then switch to this branch and run scipion update to see what happens.

**Comment by azazellochg** [comment]
Updating from devel to this branch while staying with 3.8 works. If a user wants to update python, that would have to be a new installation.

**Comment by delarosatrevin** [comment]
I would say that is reasonable to require a new installation with this Python update.

**Comment by pconesa** [comment]
Yes, but new installation means core plus all plugins. We should not require a new config nor installing new software.... it not smooth now.

**Comment by azazellochg** [comment]
I could not compile xmipp python bindings with 3.12, there are lots of errors. So, will switch to 3.11 to see if that works

**Comment by azazellochg** [comment]
> Yes, but new installation means core plus all plugins. We should not require a new config nor installing new software.... it not smooth now.

This PR does not require anything from a user, they can keep going with 3.8

1) you cannot just copy site packages between different pythons even on the same machine and assume this will work
2) "conda update python" is also not a good idea

**Comment by azazellochg** [comment]
Hi @pconesa are you able to review this? Feel free to invite more reviewers. For existing installations (python 3.8), users will have to install a new dependency: `scipion3 pip install importlib-metadata`
For new installations with python 3.12, xmipp compiles okay. I'm happy to answer any questions.

**Comment by azazellochg** [comment]
If you are still concerned about requirements change, we could at least merge other files. I can revert the requirements.

File: pyworkflow/gui/plotter.py [hunk 1]
```diff
@@ -97,7 +97,7 @@ def __init__(self, x=1, y=1, mainTitle="",
             if mainTitle:
                 self.figure.suptitle(mainTitle, fontsize=fontsize + 4)
             if windowTitle:
-                self.figure.canvas.set_window_title(windowTitle) 
+                self.figure.canvas.manager.set_window_title(windowTitle)
```
Comment (pconesa) [inline pyworkflow/gui/plotter.py:100]: Do you remember why this change?

File: pyworkflow/mapper/xmlmapper.py [hunk 2]
```diff
@@ -26,11 +26,7 @@
 # **************************************************************************
 
 from os.path import exists
-
-try:
```
Comment (pconesa) [inline pyworkflow/mapper/xmlmapper.py:30]: I think we can get rid of xmlmapper completely?

File: pyworkflow/plugin.py [hunk 3]
```diff
@@ -52,6 +49,12 @@
 from pyworkflow.template import Template
 from pyworkflow.utils import sortListByList
 
+try:
+    import importlib_metadata
+except ModuleNotFoundError:
```
Comment (pconesa) [inline pyworkflow/plugin.py:54]: Should we address this with conditional requirements?

I think import lib is backported. Not sure if we have installations in 3.6 or 3.7 though.

File: pyworkflow/gui/plotter.py (hunk 1 above)
Comment (azazellochg) [reply pyworkflow/gui/plotter.py:100]: https://stackoverflow.com/questions/5812960/change-figure-window-title-in-pylab

File: pyworkflow/mapper/xmlmapper.py (hunk 2 above)
Comment (azazellochg) [reply pyworkflow/mapper/xmlmapper.py:30]: sure, but that was not my goal in this PR :)

File: pyworkflow/plugin.py [hunk 4]
```diff
@@ -131,19 +134,17 @@ def getPlugins(cls):
     @classmethod
     def _discoverPlugins(cls):
         # Get the list of plugins registered
-        plugin_modules = []
-        for entry_point in pkg_resources.iter_entry_points('pyworkflow.plugin'):
-            plugin_modules.append(entry_point.name)
+        plugin_modules = importlib_metadata.entry_points(group='pyworkflow.plugin')
```
Comment (pconesa) [inline pyworkflow/plugin.py:137]: I was testing this this morning. My entry_points method does not have any argument ?

File: pyworkflow/plugin.py (hunk 3 above)
Comment (azazellochg) [reply pyworkflow/plugin.py:54]: https://docs.python.org/3/library/importlib.metadata.html
importlib_metadata is the backport of importlib.metadata. It does require python 3.8+, yes.

**Comment by pconesa** [comment]
I think I had the not backported or the other one. Once I pip installed pyworkfloe it installed importlib_metadata (probably homgenizing everithing).
________________________________
De: Grigory Sharov ***@***.***>
Enviado: lunes, 29 de abril de 2024 16:17
Para: scipion-em/scipion-pyworkflow ***@***.***>
Cc: Pablo Conesa ***@***.***>; Mention ***@***.***>
Asunto: Re: [scipion-em/scipion-pyworkflow] for python 3.8-3.12 (PR #402)


@azazellochg commented on this pull request.

________________________________

In pyworkflow/plugin.py<https://github.com/scipion-em/scipion-pyworkflow/pull/402#discussion_r1583169826>:

> @@ -131,19 +134,17 @@ def getPlugins(cls):
     @classmethod
     def _discoverPlugins(cls):
         # Get the list of plugins registered
-        plugin_modules = []
-        for entry_point in pkg_resources.iter_entry_points('pyworkflow.plugin'):
-            plugin_modules.append(entry_point.name)
+        plugin_modules = importlib_metadata.entry_points(group='pyworkflow.plugin')


That's strange: https://importlib-metadata.readthedocs.io/en/latest/api.html#importlib_metadata.entry_points
How did you test this? I'd like to try as well

**Comment by azazellochg** [comment]
For now, we can update just pyworkflow. pwem and app PRs will be later. I'm running devel installation with this specific branch for pw and python 3.8 to test possible bugs / incompatibilities.

**Comment by azazellochg** [comment]
I found that metadata viewer requires pillow 9.5.0 https://github.com/scipion-em/metadata-viewer/blob/061b9ef67bce620b857b6d0e705a7fcb6bc4f13b/requirements.txt#L2 
It would be nice to update it to match this PR.

**Comment by azazellochg** [comment]
Matplotlib and pillow need to be updated in pwem to match this PR

**Review by pconesa — APPROVED** [review]
